### PR TITLE
[sharded_shuffle] Shard picker

### PR DESCRIPTION
**Pull request by cryo28**
Simple implementation of shard picker (first step of  @danielmasny' [algorithm](https://docs.google.com/document/d/1WjHYdw8eVp0RoKPh7u9TUy5tTPgJH0Jsft1yGyRNtrE/edit))

### CODE REVIEW

File: src/protocol/oprf/sharded_shuffle/shard_picker.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+#[derive(Debug, Clone, Copy)]
+struct ShardInfo {
+    empty_slots: usize,
+}
+
+struct ShardPicker {
+    total_available_slots: usize,
+    shards: Box<[ShardInfo]>,
```
[the remark this answers is not in the document]
Comment (cryo28) [reply src/protocol/oprf/sharded_shuffle/shard_picker.rs:14]: I am going to extend ShardInfo with other information about Shards (i.e. dns names/identifier of respective servers etc). Will make sure to optimize it back into `Box<[usize]>` if all those other fields will turn out unnecessary.

File: src/protocol/oprf/sharded_shuffle/shard_picker.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    dist: WeightedIndex<usize>,
+}
+
+impl ShardPicker {
+    pub fn new(num_shards: usize, shard_size: usize) -> Result<Self, WeightedError> {
```
[the remark this answers is not in the document]
Comment (cryo28) [reply src/protocol/oprf/sharded_shuffle/shard_picker.rs:19]: Could you please explain to me the benefits of this optimization?

1. All operation on Vec/slices like indexed access are perfomed on usizes anyway. So, even if I  accept num_shards as a u16 it will be cast into usize on each iteration anyway.
2. If we expect the number of shards to not be greater than 64K, then is the usize->u32 optimization for shard_size really worth? It is going to save no more than 64K * (8-4) = 256Kb of memory? Is it because of something related to fitting this all into CPU caches provide the code to be in a hot path?
3. Also, provided that shard picking is going to be in a hot path, isn't it better to let CPU operate on 64bit values (equal to register sizes), rather than let it deal with 64->32 bit conversions?

Basically I'd like to learn you line of thought here.

File: src/protocol/oprf/sharded_shuffle/shard_picker.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        let shards = vec![
+            ShardInfo {
+                empty_slots: shard_size,
+            };
+            num_shards
+        ]
+        .into_boxed_slice();
+
+        let weights = shards
```
[the remark this answers is not in the document]
Comment (cryo28) [reply src/protocol/oprf/sharded_shuffle/shard_picker.rs:28]: It is unnecessary. Thanks for pointing it out

File: src/protocol/oprf/sharded_shuffle/shard_picker.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/protocol/oprf/sharded_shuffle/shard_picker.rs:19]: I am not suggesting to do this for the purpose of making code more efficient, but rather use more appropriate types to represent this information. number of shards and shard size should not be platform-specific and 8 bytes is too much for either of those arguments imo. 

It is possible and probably ok to use `usize` internally, but rejecting invalid values at compile type by requesting appropriately-sized types for function arguments saves a lot of headache in the future.

> Also, what does "infallible" mean here? Is it some Rust-specific lingo?

Infallible functions don't fail. In this case, if arguments provided to it are guaranteed to be positive non-zero integers, then it is always possible to make this struct. `NonZeroUXX` type can help here.

**Comment by danielmasny** [comment]
We are taking a different approach that allows parallelization.
